nilearn/glm/first_level/first_level.py

```diff
@@ -124,26 +153,61 @@ def run_glm(Y, X, noise_model='ar1', bins=100, n_jobs=1, verbose=0):
     # Create the model
     ols_result = OLSModel(X).fit(Y)
 
-    if noise_model == 'ar1':
-        # compute and discretize the AR1 coefs
-        ar1 = (
-            (ols_result.residuals[1:]
-             * ols_result.residuals[:-1]).sum(axis=0)
-            / (ols_result.residuals ** 2).sum(axis=0)
-        )
+    if noise_model[:2] == 'ar':
+
+        err_msg = ('AR order must be a positive integer specified as arN, '
+                   'where N is an integer. E.g. ar3. '
+                   'You provied {0}.'.format(noise_model))
+        try:
+            ar_order = int(noise_model[2:])
+        except ValueError:
+            raise ValueError(err_msg)
+
+        # compute the AR coeficents
+        ar_coef_ = [_yule_walker(ols_result.residuals[:, res].reshape(-1, 1).T,
+                                 ar_order)
+                    for res in range(ols_result.residuals.shape[1])]
```

Comment:
did you consider vectorizing this loope?  computing the cross-correlations in yule walkers should be done easily for all voxels without a for loop. Also np.linalg.solve can operate in parallel to solve all linear systems for all voxels. See:


```
In [1]: import numpy as np

In [2]: A = np.random.randn(100, 2, 2)

In [3]: b = np.random.randn(100, 2)

In [4]: x = np.linalg.solve(A, b)

In [5]: x.shape
Out[5]: (100, 2)
```

Reply:
@agramfort I'm embarrassed to admit I have failed to vectorise this loop. Are there any further hints you can provide?